### PR TITLE
feat(ui): Conditions to word-form numbers in conversations

### PR DESCRIPTION
**Pull request by UnorderedSigh**
**Feature:** Using the `&[format@condition]` syntax, you can print "twelve thousand" instead of "12,000"

## Feature Details
The `&[format@condition]` syntax allows inserting formatted conditions into conversations. This PR adds word-form numbers (like "thirty-three"). You can also use the Chicago Manual of Style or MLA Handbook style to automatically choose between numbers and words. It correctly handles negative numbers, too.

- `&[words@condition]` = convert the condition to word form (ie. thirty-three)
- `&[chicago@condition]` = convert to word form if the number suits it (according to the Chicago Manual of Style). Otherwise, use numbers (ie. 12,001)
- `&[mla@condition]` = like "chicago" but with the MLA Handbook style (which includes hybrid forms: "12.6 million")
- `&[Words@condition]`, `&[Chicago@condition]`, or `&[Mla@condition]` = convert the condition to word form and capitalize the first letter. This is intended for use at the beginning of the sentence.

```ruby
conversation "Numbers are fun!"
    action
        "number of ships" = 101
    `&[Chicago@number of ships] ships will attack you. Be ready for them!`
    # => One hundred one ships will attack you. Be ready for them!
    choice
        `Wow! They have &[chicago@number of ships] ships?`
        # => Wow! They have 101 ships?
        `A pitiful &[words@number of ships] ships won't stop the great Captain <last>!`
        # => A pitiful one hundred one ships won't stop the great Captain <last>!
        `I eat more ships than that for breakfast. Let's do this.`
```

## UI Screenshots
word-numbers

## Usage Examples
Make numbers in conversations flow better with the text.

## Testing Done
1. Put this in data/[wordformtest.txt](https://github.com/endless-sky/endless-sky/files/11571928/wordformtest.txt)
2. Visit the spaceport at Mars.
3. Play with numbers and see their representation.

### Automated Tests Added
Unit tests in `test_format.cpp`

## Performance Impact
N/A

### CODE REVIEW

**Comment by UnorderedSigh** [comment]
This is definitely one of my least interesting PRs, but someone had to implement it.

**Comment by UnorderedSigh** [comment]
Last tested and confirmed to work Mon, 29 May 2023 17:20:06 UTC

**Comment by UnorderedSigh** [comment]
I just added MLA's hybrid word/numbers (ie. 21316000 => 21.316 million)'

**Comment by UnorderedSigh** [comment]
The failure in clang is an incompatibility with its own C++ library. This may have been due to a failed software update. I'll try rerunning the tests.

```
  C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.37.32822\include\yvals_core.h(847,1): error: static assertion failed: Error in C++ Standard Library usage.
  _EMIT_STL_ERROR(STL1000, "Unexpected compiler version, expected Clang 16.0.0 or newer.");
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.37.32822\include\yvals_core.h(502,5): note: expanded from macro '_EMIT_STL_ERROR'
      static_assert(false, "Error in C++ Standard Library usage.")
      ^             ~~~~~
```

**Comment by UnorderedSigh** [comment]
Last tested and known to work Sun, 27 Aug 2023 15:46:35 UTC
(The clang error went away when I simply reran the tests.)

**Comment by UnorderedSigh** [comment]
Last tested and known to work Thu, 23 Nov 2023 00:15:24 UTC.

File: source/text/Format.cpp [hunk 1]
```diff
@@ -24,6 +24,96 @@ this program. If not, see <https://www.gnu.org/licenses/>.
...
+			const size_t BUFLEN = 100;
+			char buf[BUFLEN] = { 0 };
+			snprintf(buf, BUFLEN, "%s%.3f %s",
+				negative ? "negative " : "", above / 1000.0, WORD_NUMBERS[magnitude].first);
+			buf[BUFLEN - 1] = '\0';
+			return string(buf);
+		}
+		return string();
+	}
+
```
Comment (shitwolfymakes) [inline source/text/Format.cpp:116]: I would move this function to immediately follow `MLAForm`, as that is the function that calls this one and it logically follows for it to be after, even tho it's not being exposed by the header.

**Review by shitwolfymakes — COMMENTED** [review]
Other than the one comment I have no notes. Very happy to see test coverage included in the PR!

File: source/text/Format.cpp (hunk 1 above)
Comment (petervdmeer) [reply source/text/Format.cpp:116]: I understand your comment, but I feel that this location is also fine. (And declaring it after MLAForm would require the function to be declared separately.)

**Comment by petervdmeer** [comment]
> Very happy to see test coverage included in the PR!

Agreed!

**Pull request merged**